### ISSUE TITLE
VScode behavior is very slow while using clientExtensions

### ISSUE THREAD

**Issue by drekthral** [prisma/prisma#17843]
### Bug description

_Today I updated prisma to 4.10.0 and immediately experienced very slow VScode behavior in project where prisma is installed._

### **EDIT:**

After few hours of playing with VScode extensions, code, schema, ... I ended up with new observings.

Seems like the version is **not** causing the issue. Let me explain:

Before updating to version 4.10.0 I prepared this code for extending the prisma client:
```
.$extends({
  result: {
    iOrder: {
      orderNumber: {
        needs: { id: true, randomNumber: true },
        compute(order) {
          return +`${order.id}${order.randomNumber}`;
        }
      }
    }
  },
  model: {
    iOrder: {
      byOrderNumber: (vs: number) => {
        const str = `${vs}`;
        const split = [parseInt(str.slice(0, str.length-2)), parseInt(str.slice(str.length-2))];
        return {
          id: split[0],
          randomNumber: split[1]
        };
      }
    } satisfies Record<string, (...args: any) => Prisma.IOrderWhereInput>,
  }
});
```

Right after updating the version I extended the client with this code and added `clientExtensions` to `previewFeatures` and ran `npx prisma generate`. After that the VScode window was **very** slow (intellisense, error higlighting, ...).
For example after typing `(new PrismaClient()).` the intellisense would show up in +- 13 seconds. Because of that I wrongly blamed the version - sorry about that.

Now what I discovered:

1. I removed the extending code (see above) and the intellisense show time dropped to +- 3 seconds.
2. I removed `clientExtensions` from `previewFeatures` and the show time was almost instant (as it was). 

So it seems like that the `clientExtensions` preview feature is causing this problem.

Sorry about not posting our schema as we find it sensitive. But there are 24 models.

Btw [this](https://pcsupport.lenovo.com/us/en/products/LAPTOPS-AND-NETBOOKS/THINKPAD-T-SERIES-LAPTOPS/THINKPAD-T14-TYPE-20UD-20UE/20UE/20UES2WA00) is my machine.

### **EDIT2:**
This slow behavior is not immediate. It shows up after few minutes of working. And it seems, like `prismaTransactionClient.` takes longer to show intellisense than `prismaClient.`

### How to reproduce

1. use `clientExtensions` preview feature
2. extend the client
3. experience the slowness of VScode - for example intellisense

### Expected behavior

VScode should be quick with and without using `clientExtensions`.

### Environment & setup

- OS: Windows 11
- Database: PostgreSQL
- Node.js version: 16.15.1

### Prisma Version

```
prisma                  : 4.10.0
@prisma/client          : 4.10.0
Current platform        : windows
Query Engine (Node-API) : libquery-engine ca7fcef713137fa11029d519a9780db130cca91d (at node_modules\@prisma\engines\query_engine-windows.dll.node)
Migration Engine        : migration-engine-cli ca7fcef713137fa11029d519a9780db130cca91d (at node_modules\@prisma\engines\migration-engine-windows.exe)
Format Wasm             : @prisma/prisma-fmt-wasm 4.10.0-84.ca7fcef713137fa11029d519a9780db130cca91d
Default Engines Hash    : ca7fcef713137fa11029d519a9780db130cca91d
Studio                  : 0.481.0
Preview Features        : clientExtensions
```

([slack thread](https://prisma-company.slack.com/archives/C02PJ7JGM6K/p1676750364663279?thread_ts=1676709798.085849&cid=C02PJ7JGM6K))

**Comment by mcbethio** [comment]
Hey @drekthral, A few things.
1. You had no issues on previous versions of Prisma?
2. Can you share your schema or a reproducible case for us - That makes our life so much easier?
Thanks.

**Comment by drekthral** [comment]
> Hey @drekthral, A few things.
> 
> 1. You had no issues on previous versions of Prisma?
> 2. Can you share your schema or a reproducible case for us - That makes our life so much easier?
>    Thanks.

Managed to find more info - the issue is updated.

**Comment by millsp** [comment]
Hey @drekthral, I have tried this on a project with a huge and complex schema but could not reproduce. Could you provide us with your schema as well as your tsconfig.json, and ideally a full reproduction? That would help a lot, thank you!

**Comment by drekthral** [comment]
Hi @millsp 

tsconfig.json
```
{
  "compilerOptions": {
    "module": "commonjs",
    "noImplicitReturns": true,
    "noUnusedLocals": true,
    "outDir": "lib",
    "sourceMap": true,
    "strict": true,
    "target": "es2017",
    "esModuleInterop": true
  },
  "compileOnSave": true,
  "include": [
    "src"
  ]
}
```

Could I send you the schema privately somehow? We do not want it to be public.

**Comment by millsp** [comment]
Thanks. Yes, please send us your schema to [schemas@prisma.io](mailto:schemas@prisma.io). If needed we can also do an NDA.

**Comment by drekthral** [comment]
@millsp The schema was sent.

One more note to the issue: This slow behavior is not immediate. It shows up after few minutes of working. And it seems, like `prismaTransactionClient.` takes longer to show intellisense than `prismaClient.`

**Comment by Fiaxhs** [comment]
Adding my experience with the same issue here: prettier on save would take 4-6s, intellisense was super slow. This isn't the case when not calling `$extends` on the prisma client.

By digging a bit I saw that eslint was complaining that applying the rules was taking a long time.

With trial and error, commenting these two rules solved the problem completely.

```
    "@typescript-eslint/no-floating-promises": ["error"],
    "@typescript-eslint/no-misused-promises": ["error"],
```

**Comment by millsp** [comment]
Thanks @drekthral, we've received your schema.

**Comment by ebg1223** [comment]
Same thing happening to me. Even just turning on "clientExtensions" in the schema file makes  type resolution/completion markedly slower. Actually trying to extend makes it grind to a hault

**Comment by aniravi24** [comment]
I'm also seeing something similar with my schema (nearly opened an issue but saw this was already here). Have a few hundred tables and doing a performance trace with TypeScript and [analyze-trace](https://github.com/microsoft/typescript-analyze-trace) shows me there's several hotspots related to evaluating types that are suffixed with "Needs". Looked in the index.d.ts and saw those were all client extension related.

VSCode and the typechecker basically grind to a halt and it's difficult to work. Once I remove the reference to the client extension, everything is speedy again.

**Comment by aniravi24** [comment]
It turns out my issue might have been caused by a bad type cast in the Prisma example [here](https://github.com/prisma/prisma-client-extensions/blob/main/callback-free-itx/script.ts#L58). This type cast causes the typechecker to become much slower, and when moving this `$begin` function outside of the client extension into a separate function, it hangs the typechecker altogether (the perf trace file gets larger and larger with the typechecker never finishing, even after several minutes).

Removing client extensions does make navigating VSCode a bit smoother (and typechecking somewhat faster), but not as big of an improvement as fixing this bad type. Just thought I'd comment here in case anyone else runs into this too.

**Comment by janpio** [comment]
Awesome @aniravi24, that is a great hint for us to double check.

**Comment by anton-g** [comment]
We're seeing similar results in our project but it's not isolated to VS Code but rather all TS transpilation. Just adding an empty extension causes our TS transpile time to increase by 200-300%. Running a trace shows the first module that imports `prisma` taking 25~ *seconds* to transpile.

```ts
const prisma = new PrismaClient().$extends({
  result: {
    advert: {},
  },
})
export { prisma }
```

**Comment by millsp** [comment]
Hey @anton-g, would you be able to provide us with your schema? If ti contains sensitive information you could also send it at schemas@prisma.io.

**Comment by niba** [comment]
I have the same problem. My vscode started working very slowly and after a few days I discovered that the reason is prisma extension mechanism. I think typescript has some problems with parsing prisma extensions types. Maybe there is some recurrency in types or something like that?

**Comment by millsp** [comment]
Hey, we have refactored the code generation to switch from fully generated types to dynamic typing. This should have improved the performance drops that you have reported. I love to hear your feedback @niba @anton-g @aniravi24 @drekthral to know if things got better on your end. Thanks. The fix is available in `4.16.0`.
